File: src/libshoopdaloop/internal/midi_helpers.cpp
#include "midi_helpers.h"

uint32_t channel(const uint8_t *msg_data) { return msg_data[0] & 0x0F; }

uint32_t note(const uint8_t *msg_data) { return msg_data[1]; }

uint32_t velocity(const uint8_t *msg_data) { return msg_data[2]; }

bool is_noteOn(const uint8_t *msg_data) { return (msg_data[0] & 0xF0) == 0x90; }

bool is_noteOff(const uint8_t *msg_data) {
    return (msg_data[0] & 0xF0) == 0x80;
}

bool is_cc(const uint8_t *msg_data) { return (msg_data[0] & 0xF0) == 0xB0; }

bool is_program(const uint8_t *msg_data) {
    return (msg_data[0] & 0xF0) == 0xC0;
}

bool is_channel_pressure(const uint8_t *msg_data) {
    return (msg_data[0] & 0xF0) == 0xD0;
}

bool is_pitch_wheel(const uint8_t *msg_data) {
    return (msg_data[0] & 0xF0) == 0xE0;
}

std::optional<uint32_t> is_all_notes_off_for_channel(const uint8_t *msg_data) {
    if (is_cc(msg_data) && msg_data[1] == 123) {
        return channel(msg_data);
    }
    return std::nullopt;
}

std::optional<uint32_t> is_all_sound_off_for_channel(const uint8_t *msg_data) {
    if (is_cc(msg_data) && msg_data[1] == 120) {
        return channel(msg_data);
    }
    return std::nullopt;
}

std::vector<uint8_t> noteOn(uint8_t channel, uint8_t note, uint8_t velocity) {
     return { static_cast<unsigned char>(0x90 + channel), note, velocity };
}

std::vector<uint8_t> noteOff(uint8_t channel, uint8_t note, uint8_t velocity) {
    return { static_cast<unsigned char>(0x80 + channel), note, velocity };
}

std::vector<uint8_t> cc(uint8_t channel, uint8_t controller, uint8_t value) {
    return { static_cast<unsigned char>(0xB0 + channel), controller, value };
}

std::vector<uint8_t> programChange(uint8_t channel, uint8_t program) {
    return { static_cast<unsigned char>(0xC0 + channel), program };
}

std::vector<uint8_t> pitchWheelChange(uint8_t channel, uint16_t value) {
    uint8_t first = (uint8_t)(value & 0b1111111);
    uint8_t second = (uint8_t)((value >> 8) & 0b1111111);
    return { static_cast<unsigned char>(0xE0 + channel), first, second };
}

std::vector<uint8_t> channelPressure(uint8_t channel, uint8_t value) {
    return { static_cast<unsigned char>(0xD0 + channel), value };
}

File: src/rust/backend_rust/src/midi_helpers.rs
pub fn channel(data: &[u8]) -> u32 {
    (data[0] & 0x0F) as u32
}

pub fn note(data: &[u8]) -> u32 {
    data[1] as u32
}

pub fn velocity(data: &[u8]) -> u32 {
    data[2] as u32
}

pub fn is_note_on(data: &[u8]) -> bool {
    data[0] & 0xF0 == 0x90 && data[2] != 0
}

pub fn is_note_off(data: &[u8]) -> bool {
    data[0] & 0xF0 == 0x80 || (data[0] & 0xF0 == 0x90 && data[2] == 0)
}

pub fn is_cc(data: &[u8]) -> bool {
    data[0] & 0xF0 == 0xB0
}

pub fn is_program(data: &[u8]) -> bool {
    data[0] & 0xF0 == 0xC0
}

pub fn is_channel_pressure(data: &[u8]) -> bool {
    data[0] & 0xF0 == 0xD0
}

pub fn is_pitch_wheel(data: &[u8]) -> bool {
    data[0] & 0xF0 == 0xE0
}

pub fn is_all_notes_off_for_channel(data: &[u8]) -> Option<u32> {
    if is_cc(data) && data[1] == 123 {
        Some(channel(data))
    } else {
        None
    }
}

pub fn is_all_sound_off_for_channel(data: &[u8]) -> Option<u32> {
    if is_cc(data) && data[1] == 120 {
        Some(channel(data))
    } else {
        None
    }
}

pub fn note_on(channel: u8, note: u8, velocity: u8) -> Vec<u8> {
    vec![0x90 | channel, note, velocity]
}

pub fn note_off(channel: u8, note: u8, velocity: u8) -> Vec<u8> {
    vec![0x80 | channel, note, velocity]
}

pub fn cc(channel: u8, controller: u8, value: u8) -> Vec<u8> {
    vec![0xB0 | channel, controller, value]
}

pub fn program_change(channel: u8, program: u8) -> Vec<u8> {
    vec![0xC0 | channel, program]
}

pub fn pitch_wheel_change(channel: u8, value: u16) -> Vec<u8> {
    vec![
        0xE0 | channel,
        (value & 0x7F) as u8,
        ((value >> 7) & 0x7F) as u8,
    ]
}

pub fn channel_pressure(channel: u8, value: u8) -> Vec<u8> {
    vec![0xD0 | channel, value]
}

pub fn midi_message_len(status: u8) -> usize {
    match status & 0xF0 {
        0xC0 | 0xD0 => 2,
        _ => 3,
    }
}
